### PR TITLE
Collection Camp Authorisation Status

### PR DESCRIPTION
**Pull request by tarunnjoshi**
Remove the outcome link and change the status

Target Issue - https://github.com/coloredcow-admin/goonj-crm/issues/133 

## How to Test
1. **Planned**: Authorize any camp and go to that particular camp where you'll able to see the `Camp status field` become Planned
2. If anyone marked the camp status **Aborted**: If it's aborted then on the logistics mail the camp outcome link is not there and users are not able to fill out the camp outcome form.
3. If he selected **Aborted** then there is a field for the reason for aborting there on the form.
4. **Completed**: If outcome form for any camp is filled then the `Camp status field` become Completed

## Summary by CodeRabbit

- **New Features**
	- Enhanced event subscription management for camp status updates.
	- Conditional email content based on camp status.
	- Improved processing based on camp status for various cron functions.

- **Bug Fixes**
	- Improved processing logic to skip camps with an "aborted" status.

- **Documentation**
	- Updated method signatures to reflect new functionalities and changes.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 1]
```diff
@@ -1138,4 +1154,97 @@ public static function updateContributorCount($collectionCamp) {
...
+        }
+
+        // Update camp status to 'planned' when the camp is authorized
+        $results = EckEntity::update('Collection_Camp', TRUE)
+          ->addValue('Collection_Camp_Intent_Details.Camp_status_field', 'planned')
+          ->addWhere('id', '=', $campId)
+          ->execute();
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (tarunnjoshi) [reply wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1209]: sure

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampOutcomeService.php [hunk 2]
```diff
@@ -25,6 +25,13 @@ public static function processCampReminder($camp, $now, $from) {
     $collectionCampId = $camp['id'];
     $campCode = $camp['title'];
     $campAddress = $camp['Collection_Camp_Intent_Details.Location_Area_of_camp'];
+    $campStatus = $camp['Collection_Camp_Intent_Details.Camp_status_field'];
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampOutcomeService.php:28]: Can we please recreate this custom field and remove `_field` from its name?

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampOutcomeService.php [hunk 3]
```diff
@@ -25,6 +25,13 @@ public static function processCampReminder($camp, $now, $from) {
...
+
+    // Skip if camp status is "aborted".
+    if ($campStatus === 'aborted') {
+      \Civi::log()->info("Skipping camp ID $collectionCampId as it is marked as 'aborted'");
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampOutcomeService.php:32]: No need to log this.

Comment (pokhiii) [reply wp-content/civi-extensions/goonjcustom/Civi/CollectionCampOutcomeService.php:32]: Let's move this a where clause on the query

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 4]
```diff
@@ -276,22 +277,15 @@ public static function assignChapterGroupToIndividual(string $op, string $object
...
   public static function generateCollectionCampCode(string $op, string $objectName, $objectId, &$objectRef) {
-    if ($objectName != 'Eck_Collection_Camp') {
-      return;
-    }
+    $statusDetails = self::checkCampStatusAndIds($objectName, $objectId, $objectRef);
 
-    $newStatus = $objectRef['Collection_Camp_Core_Details.Status'] ?? '';
-
-    if (!$newStatus || !$objectId) {
+    // Ensure valid status and camp details are returned.
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:282]: these comments are not needed. the code itself speaks for it.

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 5]
```diff
@@ -1046,6 +1040,13 @@ public static function sendLogisticsEmail($collectionCamp) {
       $campAddress = $collectionCamp['Collection_Camp_Intent_Details.Location_Area_of_camp'];
       $campAttendedById = $collectionCamp['Logistics_Coordination.Camp_to_be_attended_by'];
       $logisticEmailSent = $collectionCamp['Logistics_Coordination.Email_Sent'];
+      $campStatus = $collectionCamp['Collection_Camp_Intent_Details.Camp_status_field'];
+
+      // Skip if camp status is "aborted".
+      if ($campStatus === 'aborted') {
+        \Civi::log()->info("Skipping camp ID $collectionCampId as it is marked as 'aborted'");
+        return FALSE;
+      }
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1049]: Let's move this check to API query where clause

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 6]
```diff
@@ -1143,4 +1145,133 @@ public static function updateContributorCount($collectionCamp) {
...
+   *   The name of the object.
+   * @param int $objectId
+   *   The unique identifier for the object.
+   * @param object $objectRef
+   *   The reference to the object.
+   */
+  public static function updateCampStatusAfterAuth(string $op, string $objectName, $objectId, &$objectRef) {
+    $statusDetails = self::checkCampStatusAndIds($objectName, $objectId, $objectRef);
+
+    // Ensure valid status and camp details are returned.
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1163]: comment not needed as it does not add more value to readability

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 7]
```diff
@@ -1143,4 +1145,133 @@ public static function updateContributorCount($collectionCamp) {
...
+    if (!$statusDetails) {
+      return;
+    }
+
+    $newStatus = $statusDetails['newStatus'];
+    $currentStatus = $statusDetails['currentStatus'];
+
+    // Check for status change.
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1171]: Not needed

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 8]
```diff
@@ -1143,4 +1145,133 @@ public static function updateContributorCount($collectionCamp) {
...
+    if ($currentStatus !== $newStatus) {
+      if ($newStatus === 'authorized') {
+        // Access the subtype.
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1174]: absolutely not needed

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 9]
```diff
@@ -1143,4 +1145,133 @@ public static function updateContributorCount($collectionCamp) {
...
+        $subtypeId = $objectRef['subtype'] ?? NULL;
+        if ($subtypeId === NULL) {
+          return;
+        }
+
+        // Access the id within the decoded data.
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1180]: remove

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php [hunk 10]
```diff
@@ -1143,4 +1145,133 @@ public static function updateContributorCount($collectionCamp) {
...
+        $campId = $objectRef['id'] ?? NULL;
+        if ($campId === NULL) {
+          return;
+        }
+
+        // Update camp status to 'planned' when the camp is authorized.
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampService.php:1186]: remove

File: wp-content/civi-extensions/goonjcustom/Civi/CollectionCampVolunteerFeedbackService.php [hunk 11]
```diff
@@ -34,6 +34,13 @@ public static function processVolunteerFeedbackReminder($camp, $now, $from) {
     $endDate = new \DateTime($camp['Collection_Camp_Intent_Details.End_Date']);
     $collectionCampId = $camp['id'];
     $campAddress = $camp['Collection_Camp_Intent_Details.Location_Area_of_camp'];
+    $campStatus = $camp['Collection_Camp_Intent_Details.Camp_status_field'];
+
+    // Skip if camp status is "aborted".
+    if ($campStatus === 'aborted') {
+      \Civi::log()->info("Skipping camp ID $collectionCampId as it is marked as 'aborted'");
+      return FALSE;
+    }
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/CollectionCampVolunteerFeedbackService.php:43]: move to api query if possible

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/VolunteerFeedbackCollectionCampCron.php [hunk 12]
```diff
@@ -66,6 +66,13 @@ function civicrm_api3_goonjcustom_volunteer_feedback_collection_camp_cron($param
       $feedbackEmailSent = $camp['Logistics_Coordination.Feedback_Email_Sent'];
       $initiatorId = $camp['Collection_Camp_Core_Details.Contact_Id'];
       $campAddress = $camp['Collection_Camp_Intent_Details.Location_Area_of_camp'];
+      $campStatus = $camp['Collection_Camp_Intent_Details.Camp_status_field'];
+
+      // Skip if camp status is "aborted".
+      if ($campStatus === 'aborted') {
+        \Civi::log()->info("Skipping camp ID $collectionCampId as it is marked as 'aborted'");
+        continue;
+      }
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/VolunteerFeedbackCollectionCampCron.php:75]: where clause in api query

**Review by pokhiii — CHANGES_REQUESTED** [review]

**Review by pokhiii — APPROVED** [review]

**Pull request merged**